src/Analyzers/CSharp/Analyzers/UseParameterNullChecking/CSharpUseParameterNullCheckingDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,253 @@
+                                if (referenceEqualsMethod == null || !referenceEqualsMethod.Equals(semanticModel.GetSymbolInfo(receiver, cancellationToken).Symbol))
+                                {
+                                    continue;
+                                }
+
+                                left = arguments[0].Expression;
+                                right = arguments[1].Expression;
+                                break;
+
+                            default:
+                                continue;
+                        }
+
+                        if (!AreOperandsApplicable(left, right, out parameter)
+                            && !AreOperandsApplicable(right, left, out parameter))
+                        {
+                            continue;
+                        }
+
+                        var throwStatement = ifStatement.Statement switch
+                        {
+                            ThrowStatementSyntax @throw => @throw,
+                            BlockSyntax { Statements: { Count: 1 } statements } => statements[0] as ThrowStatementSyntax,
+                            _ => null
+                        };
+
+                        if (throwStatement is null
+                            || throwStatement.Expression is not ObjectCreationExpressionSyntax thrownInIf
+                            || !IsConstructorApplicable(thrownInIf, parameter))
+                        {
+                            continue;
+                        }
+
+                        break;
+
+                    // this.field = param ?? throw new ArgumentNullException(nameof(param));
+                    case ExpressionStatementSyntax
+                    {
+                        Expression: AssignmentExpressionSyntax
+                        {
+                            Right: BinaryExpressionSyntax
+                            {
+                                OperatorToken.RawKind: (int)SyntaxKind.QuestionQuestionToken,
+                                Left: ExpressionSyntax maybeParameter,
+                                Right: ThrowExpressionSyntax { Expression: ObjectCreationExpressionSyntax thrownInNullCoalescing }
+                            }
+                        }
+                    }:
+                        if (!IsParameter(maybeParameter, out parameter) || !IsConstructorApplicable(thrownInNullCoalescing, parameter))
+                        {
+                            continue;
+                        }
+
+                        break;
+
+                    default:
+                        continue;
+                }
+
+                context.ReportDiagnostic(DiagnosticHelper.Create(
+                    Descriptor,
+                    statement.GetLocation(),
+                    option.Notification.Severity,
+                    additionalLocations: new[] { parameter.Locations[0] },
```

Comment:
Can a parameter have multiple locations? *For example*, for partial methods, does the compiler produce one or two `ParameterSymbol`s?

Reply:
IIRC, A partial method implementation is its own symbol and has its own distinct parameter symbols. It's not a case where the parameter symbol is shared between the two parts and has multiple locations, for example. I'll make sure we have the test, though.